lib/pmdtester/runner.rb

```diff
@@ -49,19 +51,63 @@ def run_local_mode
     end
 
     def run_online_mode
-      # TODO
+      puts "Mode: #{@options.mode}"
+      check_option(ONLINE, 'base branch name', @options.base_branch)
+      check_option(ONLINE, 'patch branch name', @options.patch_branch)
+
+      baseline_path = download_baseline(@options.base_branch)
+
+      # patch branch build pmd reports with same configuration as base branch
+      config_path = "#{baseline_path}/config.xml"
+
+      # patch branch build pmd report with same list of projects as base branch
+      project_list = "#{baseline_path}/project-list.xml"
+      get_projects(project_list)
+
+      PmdReportBuilder
+        .new(config_path, @projects, @options.local_git_repo, @options.patch_branch)
+        .build
+
+      build_html_reports
+    end
+
+    def download_baseline(branch_name)
+      branch_name = branch_name.delete('/')
+      zip_filename = "#{branch_name}-baseline.zip"
+      target_path = 'target/reports'
+      FileUtils.mkdir_p(target_path) unless File.directory?(target_path)
+
+      url = get_baseline_url(zip_filename)
+      wget_cmd = "wget #{url}"
+      unzip_cmd = "unzip -qo #{zip_filename}"
+
+      Dir.chdir(target_path) do
+        Cmd.execute(wget_cmd) unless File.exist?(zip_filename)
+        Cmd.execute(unzip_cmd)
+      end
+
+      "#{target_path}/#{branch_name}"
+    end
+
+    def get_baseline_url(zip_filename)
+      "https://sourceforge.net/projects/pmd/files/pmd-regression-tester/#{zip_filename}"
     end
 
     def run_single_mode
       puts "Mode: #{@options.mode}"
       check_option(SINGLE, 'patch branch name', @options.patch_branch)
       check_option(SINGLE, 'patch branch config path', @options.patch_config)
+      check_option(SINGLE, 'list of projects file path', @options.project_list)
 
-      PmdReportBuilder
-        .new(@options.patch_config, @projects, @options.local_git_repo, @options.patch_branch)
-        .build
+      get_projects(@options.project_list) unless @options.nil?
+      branch_details = PmdReportBuilder
+                       .new(@options.patch_config, @projects,
+                            @options.local_git_repo, @options.patch_branch)
+                       .build
+      # copy list of projects file to the patch baseline
+      FileUtils.cp(@options.project_list, branch_details.target_branch_project_list_path)
 
-      build_html_reports
+      build_html_reports unless @options.html_flag
```

Comment:
I'm thinking about your new html flag.... and why we need this...
I guess, if we can only call build_html_reports if we have both a base and patch report available. So, if we run single mode with a freshly installed tool (nothing in target yet), then we have no base report and thus can't create the diff report.

Shouldn't we just remove this flag and just never execute build_html_reports when running in single mode?

Reply:
Single mode is not just for baseline in my imagination.
It can also be used to evaluate the performance of new rules. For example, patch config contains only one new rule. Then we can see how this new rule behaves on all standard projects.

Reply:
Understood, yes, that makes sense.